accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,194 @@
+library that turns on platform-level preview features requires its consumer to
+turn them on as well.
+
+Since library-level preview features cannot change how the .NET runtime itself
+operates, we don't have that same level of concern. In fact, we believe the
+viral nature is also undesirable.
+
+## Scenarios and User Experience
+
+## Requirements
+
+### Goals
+
+* Don't change the guardrails for platform-level preview features:
+    - They require opt-in (without, the build fails with an error)
+    - They require to be viral (i.e. opt-in is also required for indirect
+      dependencies relying on preview features)
+    - The opt-in can be expressed centrally in the project file
+    - They require SDK and framework to match
+
+* Offer library-level preview features with a different set of guardrails:
+    - They require opt-in (without, the build fails with an error)
+    - The opt-in isn't viral and can be an implementation detail
+    - The opt-in is centralized, but it's specific to each preview feature
+    - Using them poses no requirements on the SDK version
+
+### Non-Goals
+
+* Relaxing or extending the guardrails for platform-level preview features
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+    - Core libraries
+    - ASP.NET
+    - EF
+    - Azure SDK
+    - R9
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Attribute
+
+We'll create a new attribute, separate from the existing `System.Runtime.Versioning.RequiresPreviewFeaturesAttribute`:
+
+```C#
+namespace System.Diagnostics.CodeAnalysis;
+
+[AttributeUsage(AttributeTargets.Assembly |
+                AttributeTargets.Module |
+                AttributeTargets.Class |
+                AttributeTargets.Struct |
+                AttributeTargets.Enum |
+                AttributeTargets.Constructor |
+                AttributeTargets.Method |
+                AttributeTargets.Property |
+                AttributeTargets.Field |
+                AttributeTargets.Event |
+                AttributeTargets.Interface |
+                AttributeTargets.Delegate, Inherited = false)]
+public sealed class PreviewAttribute : Attribute
+{
+    public PreviewAttribute(string diagnosticId);
+    public string DiagnosticId { get; }
+    public string? UrlFormat { get; set; }
+}
+```
+
+### Compiler Behavior
+
+The compiler will raise a diagnostic when a preview API is used, using supplied
+diagnostic ID. The severity is always error.
+
+The semantics are identical to how obsolete is tracked, except that it will
+always raise a diagnostic regardless of whether the call site is marked as
+preview or not. There is also no special treatment when both caller and callee
+are in the same assembly -- any use will generate a diagnostic and the caller is
+expected to suppress it, with the usual means (i.e. `#pragma` or project-wide
+`NoWarn`).
```

Comment:
@jaredpar @333fred If this becomes a compiler feature, what would be our downlevel story? For analzyers, downlevel consumers can just reference the analyzer as a package. AFAIK that's possible for the compiler too but I don't think this will play nicely in all the cases, right (like the IDE won't use that compiler).

Does this mean that for downlevel consumers we'd still need an analyzer? If so, then that's unfortunate because it means the suppression model for preview APIs would be either on/off for the entire project or use-site suppression (basically how obsoletion worked before we added the diagnostic id).

Any ideas?

Reply:
If this is a compiler feature, it will necessarily prevent downlevel SDKs from benefiting from the feature, _unless_ the user referenced `Microsoft.NET.Compilers.Toolset` from their project. We could poison-pill the apis with a `CompilerFeatureRequired` attribute on it as well, but that is also a fairly new compiler and I wouldn't expect the .NET 6 SDK to block access to the API.

Reply:
> For analzyers, downlevel consumers can just reference the analyzer as a package. 

Only so far back though that the compiler supports the API used by the analyzer. That is a way that we already put limits on how old of a compiler can be used.